src/main/java/org/codehaus/groovy/runtime/typehandling/NumberMath.java

```diff
@@ -168,11 +168,26 @@ public static boolean isBigInteger(Number number) {
     }
 
     public static BigDecimal toBigDecimal(Number n) {
-        return (n instanceof BigDecimal ? (BigDecimal) n : new BigDecimal(n.toString()));
+        if (n instanceof BigDecimal) {
+            return (BigDecimal) n;
+        }
+        if (n instanceof BigInteger) {
+            return new BigDecimal((BigInteger) n);
+        }
+        if (n instanceof Byte || n instanceof Short || n instanceof Integer || n instanceof Long) {
```

Comment:
I suggest to put Integer and Long at the beginning to avoid checks as possible as we could, because we use Integer and Long more often in the life 😉